core/preconditioner/jacobi.cpp

```diff
@@ -31,16 +31,16 @@ ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 ******************************<GINKGO LICENSE>*******************************/
 
-#include "core/preconditioner/jacobi.hpp"
+#include <ginkgo/core/preconditioner/jacobi.hpp>
 
 
-#include "core/base/exception_helpers.hpp"
-#include "core/base/executor.hpp"
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/executor.hpp>
+#include <ginkgo/core/base/math.hpp>
+#include <ginkgo/core/base/utils.hpp>
+#include <ginkgo/core/matrix/csr.hpp>
+#include <ginkgo/core/matrix/dense.hpp>
 #include "core/base/extended_float.hpp"
```

Comment:
Maybe we should split the block of public headers from the block of private ones?
We are already doing the same thing for [third-party headers](https://github.com/ginkgo-project/ginkgo/wiki/Contributing-guidelines#include-statement-grouping).

Reply:
Probably a good idea, however, I did not include it in my search and replace script. Do you know how to do it in a automated way (doing this by hand is very tough)?

Reply:
Looks like `clang-format` already helped a bit and grouped the blocks so they first include the system-style includes, followed by the local ones.
So something like (from the top of my head - not tested) the following should work:

```sh
find -type f -not -path "*/.git/*" -exec \
    perl -i -pe "s/(#include<ginkgo/.*)\n(#include \")/\1\n\n\n\2/g;" {} +
```

Reply:
I have tried many different things, but I can't get it to work...
As far as I know, this should work:
```bash
perl -i -pe 's/(#include <ginkgo\/.*>)\n(#include ".*")/\1\n\n\n\2/s'
```
But nothing is changed when I execute it...

Reply:
Could you help me figure out why it is not working?

As soon as anything goes beyond the line-break in the first search, nothing matches anymore. I changed the 'g' at the end to an 's', so everything is considered as one line, but it is still not working.

Reply:
Here is some piece of awk which could solve the problem. You will have to combine it with the find command in some way (either using `-exec` or with a for loop).

```bash
awk '
{
  if (/#include <ginkgo/)
  {
		l=$0
		print
		next
	}
	if ($0 ~ /#include "/ && done == 0)
	{
		print "\n"
		print $0
		done=1
	}
	else { print }
}' core/preconditioner/jacobi.cpp
```

Reply:
It's because perl reads line-by-line by default, so even though you regex matches multiple lines, it never gets multiple lines, see also: https://stackoverflow.com/a/9671078/9385966

This should do the trick:

```sh
perl -0777 -i -pe 's/(#include <.*)\n(#include ")/\1\n\n\n\2/g'
```

Reply:
@tcojean Thank you for your suggestion, it works nicely. I have never heard about awk up until now to be honest.

@gflegar Thank you as well, it also works as it should. I guess my google search was flawed, I never came across this post.